### PR TITLE
treewide: use finalAttrs.finalPackage in testers.testVersion

### PR DESCRIPTION
**Pull request by pbsds**
## Description of changes

candidates were located using `rg -F 'testers.testVersion { package =' | grep -v finalAttrs | cut -d: -f1 | xargs grep -E '(buildRustPackage|buildGoModule)' -L | xargs grep mkDerivation -l`

There are surprisingly many examples of `finalAttrs` migrations that don't change `testers.testVersion`

should be 0 rebuilds

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by JohnRTitor** [comment]
I wonder how we can handle cases like these where we have to use this `rec`ursive technique.

https://github.com/NixOS/nixpkgs/blob/a2cad4db0b6d352c0bc484f82e48c445f8092778/pkgs/by-name/ze/zed-editor/package.nix#L248

File: pkgs/development/libraries/magic-enum/default.nix
Comment (JohnRTitor) [inline pkgs/development/libraries/magic-enum/default.nix:32]: Removed by accident?

**Comment by r-vdp** [comment]
There's also `versionCheckHook` which you can simply add to `nativeBuildInputs`.
I think we could consider deprecating `testers.testVersion` in favour of that.

**Comment by pbsds** [comment]
`versionCheckHook` turns a sanity/sniff test into a build failure, a tradeoff i believe the package maintainer should get to decide. It is likely break if a package switches to the 0-unstable-yyyy-mm-dd version schema.

**Review by philiptaron — APPROVED** [review]
Smörgåsbord of derivations here.

**Comment by philiptaron** [comment]
Will merge (or you can self-merge) after the merge conflict is fixed, Peder.

**Review by philiptaron — APPROVED** [review]

**Pull request merged**
